Nav2 toolkit

pranavk-2003 opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #5070  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Turtlebot gazebo simulation.|
| Does this PR contain AI generated software? | Somewhat |

---

## Description of testing performed

  Functional tests were conducted on turtlebot3 and are working as expected


## Description of contribution in a few bullet points

- This Contribution adds a new package named `nav2_toolkit` which has QOL features, which can be handy while working with Nav2 .

## Description of documentation updates required from your changes

* Addition of 3 new parameterauto_start_saving, auto_restore_pose, save_interval_sec, so need to add that to nav2_params and documentation page


## Description of how this change was tested

All the changes were tested multiple times in simulation over a course of 3 weeks.

---

## Future work that may be required in bullet points

In total , this package will contain 3 tools i.e

- [x] pose_saver & pose_restore with added launch file in nav2_bringup.
- [ ] Relocation of BaseFootprintPublisher.
- [ ] A node which can pause and resume nav2 goals.




#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

SteveMacenski commented on nav2_toolkit/package.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0"?>
+<package format="3">
+  <name>nav2_toolkit</name>
+  <version>0.1.0</version>
```
Please match the version with the other Nav2 packages.

Also, add this package to the `navigation2/package.xml` as a dependency to link this as part of the metapackage

SteveMacenski commented on nav2_toolkit/package.xml:
```diff
@@ -0,0 +1,22 @@
+<package format="3">
+  <name>nav2_toolkit</name>
+  <version>0.1.0</version>
+  <description>Pose persistence utility node for AMCL in ROS 2</description>
```
I think this needs to be updated as a set of tools for Nav2 applications

SteveMacenski commented on nav2_toolkit/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+
+include_directories(include)
+
+
```
Use the `nav2_package()` macro from the nav2_common package

SteveMacenski commented on nav2_toolkit/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+target_link_libraries(pose_saver_node yaml-cpp)
+
+
+ament_target_dependencies(pose_saver_node
```
Checkout the new CMake style in other packages, we shouldn't be doing this anymore

SteveMacenski commented on nav2_toolkit/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+find_package(ament_index_cpp REQUIRED)
+
+include_directories(include)
+
```
(remove extra spaces)

SteveMacenski commented on nav2_toolkit/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+project(nav2_toolkit)
+
+# C++ standard
+set(CMAKE_CXX_STANDARD 17)
```
Remove, the Nav2 package macro will set this and other build flags

SteveMacenski commented on nav2_bringup/params/nav2_params.yaml:
```diff
@@ -79,7 +79,6 @@ bt_navigator:
       - nav_to_pose
```
Please fix this file, looks like an error in rebasing which removed some recent changes for new fetaures. You may want to pull in main or rebase yourself if you haven't recently

pranavk-2003 replied:
I pulled the recent changes from main, but this still looks to be there.

pranavk-2003 replied:
Hey i am also a bit confused about what to do with this.Should it be there or not?

SteveMacenski replied:
You shouldn't be modifying this file, revert these changes please

SteveMacenski commented on nav2_toolkit/README.md:
```diff
@@ -0,0 +1,12 @@
+# nav2_toolkit
+
+A modular ROS 2 node for persisting and restoring the robot's last known pose using the `/amcl_pose` topic. This is useful in applications where the robot is never manually moved (e.g. warehouse robots), and it enables automatic re-localization.
```
This toolkit has a number of nodes, not just one. This description should be describing the package and its contents. 

You can later describe each node (probably good) but it should be in a subsection for each node after the general package description

SteveMacenski replied:
@pranavk-2003 this wasnt completed

SteveMacenski commented on nav2_toolkit/config/last_known_pose.yaml:
```diff
@@ -0,0 +1,10 @@
+pose:
```
Why is this file included? Remove please

SteveMacenski replied:
Also, this should be the stamped version, we are going to want to also store the header frame ID and the time it was saved. We probably want to override the timestamp later, but it would be good to know when the last one was from.

SteveMacenski commented on nav2_bringup/launch/pose_persistent_launch.py:
```diff
@@ -0,0 +1,73 @@
+# Copyright (C) 2024 Open Source Robotics Foundation
```
This should be copyrighted to you

Also, lets rename to `nav2_toolkit_launch.py` and include any/all relevant toolkit nodes in this one. I think of this as a launch file for all the extra toolkit nodes that can be modified, disabled, etc as applications desire

pranavk-2003 replied:
> This should be copyrighted to you

By this do you mean?
# Copyright (C) 2025 Pranav Kolekar?
It shall have my name or ?

SteveMacenski replied:
Yes, or your company's name if you're doing this on company time / resources. Otherwise, you personally is fine too!

SteveMacenski commented on nav2_bringup/launch/pose_persistent_launch.py:
```diff
@@ -0,0 +1,73 @@
+
+    declare_pose_file_path_cmd = DeclareLaunchArgument(
+    'pose_file_path', 
+    default_value=os.path.join(
```
The filepath should be something set in the configuration of the node and should use some directory that the user sets up (`/home/`, `/tmp`, etc). Using a file in the shared install space of a package is very odd and we shouldn't be modifying the source's install. It should be in another location on the filesystem not related to the source / installed software to persist updates.

SteveMacenski commented on nav2_bringup/launch/pose_persistent_launch.py:
```diff
@@ -0,0 +1,73 @@
+    bringup_dir = get_package_share_directory('nav2_bringup')
+
+    # Declare Launch Arguments
+    declare_auto_start_saving_cmd = DeclareLaunchArgument(
```
I also don't see where any of these declared arguments are actually used

SteveMacenski commented on nav2_toolkit/launch/pose_saver.launch.py:
```diff
@@ -0,0 +1,25 @@
+
+def generate_launch_description():
+    return LaunchDescription([
+        DeclareLaunchArgument('auto_start_saving', default_value='true'),
```
If declared here, it doesn't need to be declared in the `nav2_bringup` file as well.

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+
+}  // namespace nav2_toolkit
+
+int main(int argc, char **argv)
```
Move the main into another file so that this node can be used as a component. 

Also, make this node a component node :-)

SteveMacenski commented on nav2_toolkit/include/nav2_toolkit/pose_saver_node.hpp:
```diff
@@ -0,0 +1,76 @@
+namespace nav2_toolkit
+{
+
+class PoseSaverNode : public rclcpp::Node
```
Class and metrics need doxygen documentation. See other packages for reference :-)

SteveMacenski replied:
Also remove any commented out values that are unused.

SteveMacenski commented on nav2_toolkit/include/nav2_toolkit/pose_saver_node.hpp:
```diff
@@ -0,0 +1,76 @@
+
+  // === Publisher ===
+  rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr initial_pose_pub_;
+  
```
Remove excessive spaces

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+  return msg;
+}
+
+std::string PoseSaverNode::get_package_config_path()
```
Remove, we shouldn't use this as other comments suggest

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+  YAML::Node node = YAML::LoadFile(filepath);
+  geometry_msgs::msg::PoseWithCovarianceStamped msg;
+  msg.header.stamp = this->now();
+  msg.header.frame_id = "map";
```
Read from file instead

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+  this->declare_parameter<double>("save_interval_sec", 5.0);
+  this->declare_parameter<std::string>(
+    "pose_file_path", 
+    ament_index_cpp::get_package_share_directory("nav2_toolkit") + "/config/last_known_pose.yaml");
```
Set default for `/tmp`

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+    "/amcl_pose", 10,
+    std::bind(&PoseSaverNode::pose_callback, this, std::placeholders::_1));
+
+  initial_pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>("/initialpose", 10);
```
Shouldn't we use the Service version of this so we get an ACK that it was successfully set? 

https://github.com/ros-navigation/navigation2/blob/main/nav2_amcl/src/amcl_node.cpp#L1573-L1577

SteveMacenski replied:
Rather than publishing the initial pose whereas you do not know if it was received, using a service means you know it was received properly due to getting a response. If none is given, then you can error handle

pranavk-2003 replied:
Hey Steve , there seems to have a confusion.What i understood what you mean be ,as i checked here https://github.com/ros-navigation/navigation2/blob/main/nav2_msgs/srv/SetInitialPose.srv and also in nav2 docs ,it looks like theres not a response for this service, please assist me more on this if i am catching something wrong

SteveMacenski replied:
There's not a response field to the service definition, but the service client receives a response from the service that it was completed successfully or not (or timed out trying because the service doesn't exist, etc).

pranavk-2003 replied:
I am trying to accomplish this but cant get it working, can you point me towards a resource or a codeblock where this is being used, it would be really helpful for me

SteveMacenski replied:
You should be able to use the [Nav2 Utils Service Client](https://github.com/ros-navigation/navigation2/blob/main/nav2_util/include/nav2_util/service_client.hpp#L31) templated by the set initial pose srv. Call `invoke` with a reasonable timeout (5s or so). If that call throws an exception it means it timed out or failed to call. If its returns a result of any kind, then it was called properly and ACKed.

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+  timer_->cancel();  // Disabled until service starts
+
+  sub_ = this->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+    "/amcl_pose", 10,
```
For all topics in pub/sub/services: remove the `/` which sets it to be the global namespace. That doesn't allow this to work with namespaced systems

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+
+void PoseSaverNode::pose_callback(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+  last_pose_ = msg;
```
Store a `geometry_msgs::msg::PoseWithCovarianceStamped` and use `*msg;` instead. Gets around potential middleware issues of reusing memory (which has happened before)

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+  if (!saving_active_ || !last_pose_) return;
+
+  try {
+    write_pose_to_file("/tmp/pose_saver.yaml");
```
Save only in the set filepath param location, not a hardcoded location and not in `get_package_config_path`

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+  const std::shared_ptr<std_srvs::srv::Trigger::Request> req,
+  std::shared_ptr<std_srvs::srv::Trigger::Response> res)
+{
+  (void)req;
```
You can just remove `req` to be `const std::shared_ptr<std_srvs::srv::Trigger::Request> /*req*/` to get around that error instead of using `(void)req;`. Update for all services

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+    RCLCPP_INFO(this->get_logger(), "AMCL reconnected! Publishing last known pose.");
+    try {
+      auto pose_msg = read_pose_from_file(pose_file_path_);
+      rclcpp::sleep_for(500ms);
```
Why sleep?

pranavk-2003 replied:
i used it for some debugging purpose, now not needed, removed it

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+
+  std::rename(tmp_path.c_str(), filepath.c_str());
+}
+void PoseSaverNode::amcl_monitor_callback()
```
Can you explain this function? Is this based on an actual issue you've noticed with AMCL crashing?

pranavk-2003 replied:
no issues with amcl crashing,The function monitors whether amcl is alive, if the subscriber count goes to zero that means amcl has stopped or crashed.

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+  }
+}
+
+void PoseSaverNode::reset_pose_file_cb(
```
Why would someone do this? I don't see an application :upside_down_face:

pranavk-2003 replied:
yea...i get it , its just redundant, removed it hence

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+  (void)req;
+  try {
+    auto pose_msg = read_pose_from_file(pose_file_path_);
+    rclcpp::sleep_for(500ms);
```
Sleep?

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+    std::bind(&PoseSaverNode::reset_pose_file_cb, this, std::placeholders::_1, std::placeholders::_2));
+
+  if (auto_start) {
+    saving_active_ = true;
```
Is `saving_active_` needed? If the timer is running shouldn't it always be true? Every time its toggled it should be stopped.

pranavk-2003 replied:
True,but i feel its better having a separate flag which keeps track of it.

SteveMacenski replied:
It doesn't, please remove. Its another piece of state that could be mishandled and cause bugs

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+    RCLCPP_INFO(this->get_logger(), "Auto-restore enabled. Waiting for AMCL...");
+  
+    // Wait for sim time to become active
+    if (this->get_parameter_or("use_sim_time", false)) {
```
Huh?

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+  try {
+    auto pose_msg = read_pose_from_file(pose_file_path_);
+    rclcpp::sleep_for(500ms);
+    initial_pose_pub_->publish(pose_msg);
```
Should we check if anyone is listening to this topic and log if there's no subscriptions?

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,296 @@
+      }
+    }
+  
+    // Wait for AMCL to subscribe to /initialpose
```
I think if we abstract out the functionality of `restore_service_cb` into a new method, that method could just be called here and called in the service callback to share this capability.

pranavk-2003 replied:
Yes good suggestion! Made it.

SteveMacenski replied:
I dont see it being used in the startup

SteveMacenski reviewed (commented):
Good first stab!

Maybe lets wait until some more of this is fixed, but eventually we'll also need to add a configuration guide entry for this package in the docs.nav2.org and a migration guide entry to let folks know about the new features!

Anywhere that "AMCL" in mentioned in logs or naming, we should remove that and generally just talk about the pose publisher (which as an implementation detail is just AMCL for now). That way its not confusing when folks are using something that isn't AMCL, like SLAM Toolbox or another solution like VSLAM.

SteveMacenski commented:
@pranavk-2003 let me know when you want me to take a look again!

pranavk-2003 commented:
> @pranavk-2003 let me know when you want me to take a look again! I see you pushed some changes but a number of items are not yet addressed 😄

I'll let you know soon , got busy with university stuff : -| .the few pushed stuff i still have to test.

pranavk-2003 commented:
Hey @SteveMacenski ,I think i am done with the code updates , you can now have a look again throughout.The doxygen docs is in progress. : )

SteveMacenski commented:
@pranavk-2003 open the failing jobs and fix the issues they lay out :-) It looks like alot of static code quality issues in linting

SteveMacenski commented on nav2_toolkit/README.md:
```diff
@@ -0,0 +1,80 @@
+| `/localise_at_last_known_position` | Publishes last saved pose to `/initialpose` |
+| `/reset_last_known_pose` | Sets all pose values in the YAML file to zero |
+
+Call a service like this:
```
Remove, these are implicit

SteveMacenski commented on nav2_toolkit/README.md:
```diff
@@ -0,0 +1,80 @@
+
+## File Save Locations
+
+When active, poses are saved to:
```
The same location should be a parameter, not hardcoded

SteveMacenski commented on nav2_toolkit/README.md:
```diff
@@ -0,0 +1,80 @@
+
+All writes are performed atomically using a `.tmp` file + rename to prevent corruption.
+
+## License
```
Remove, the license should be in this project for them to find themselves :-)

SteveMacenski commented on nav2_toolkit/README.md:
```diff
@@ -0,0 +1,80 @@
+
+This project is licensed under the [Apache License 2.0](LICENSE).
+
+## Contributing
```
Remove, this should be as part of the Nav2 general contribution instructions

SteveMacenski commented on nav2_toolkit/README.md:
```diff
@@ -0,0 +1,80 @@
+
+Please format code using ROS 2 C++ guidelines and keep commits clean.
+
+## Compatibility
```
Remove

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,251 @@
+
+  namespace fs = std::filesystem;
+  const fs::path final_path(final_path_str);
+  const fs::path tmp_path = fs::path("/tmp") / ("pose_saver_" + std::to_string(std::hash<std::string>{}(final_path_str)) + ".tmp");
```
Remove, nothing should be hardcoded, this should be a parameterized location.

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,251 @@
+
+void PoseSaverNode::pose_callback(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+  last_pose_ = std::make_shared<geometry_msgs::msg::PoseWithCovarianceStamped>(*msg);
```
No, just store the `*msg` object itself

SteveMacenski reviewed (changes requested):
Needs test coverage as well!

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,251 @@
+  if (!saving_active_ || !last_pose_) return;
+
+  try {
+    write_pose_to_file(pose_file_path_);
```
Why not write to file when AMCL publishes to the pose topic? We could always throttle that to store every `N` seconds in case it is a fast publishing topic like from VSLAM, but then we could remove this timer entirely.

SteveMacenski commented on nav2_toolkit/src/pose_saver_node.cpp:
```diff
@@ -0,0 +1,251 @@
+
+void PoseSaverNode::pose_callback(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+  last_pose_ = std::make_shared<geometry_msgs::msg::PoseWithCovarianceStamped>(*msg);
```
Ex check if we should write by the write frequency when we get a pose here and then write it in this callback if `saving_active_`

pranavk-2003 replied:
So what i understood from this is, when everytime amcl publishes a pose, it should be written to the file , instead of every N th second.But the question i have is , how would we decide when to write using timer and when to write with amcl publish update?

SteveMacenski replied:
> So what i understood from this is, when everytime amcl publishes a pose, it should be written to the file , instead of every N th second.

Each time this callback is triggered, check against the last write time (variable to store) and write if greater than the maximum write frequency. Else, do nothing. That way if this topic publishes very fast, like in VSLAM, we're not doing a ton of IO at 100hz and instead writing to files every 1hz or less to get the general pose of the robot. 

```
rclcpp::Time time_now = now();
if (time_now - last_write_time_ > write_duration_)
{
  last_write_time_ = time_now;
  // do writing operation
}
```

pranavk-2003 commented:
> Needs test coverage as well!

Hey okay, so what shall it be?test as in demos for using the tool?

pranavk-2003 commented:
> @pranavk-2003 open the failing jobs and fix the issues they lay out :-) It looks like alot of static code quality issues in linting

Okay i understood how to solve those errors, i would solve these after all the code adjustments are done , as the code adjustments would introduce more of them

SteveMacenski commented:
> Hey okay, so what shall it be?test as in demos for using the tool?

You should create a unit test that runs the object and then publishes some data to the topic its listening to. Check after awhile that there's a file at the requested location and the data is correct. 

Then run another tests with autostart on and test that it correctly reads from that file and publishes / calls the service to set the pose correctly (the test object should expose the service, so nothing else should be running).

Then another test should be with autostart / auto-store off and call the services to start / stop / reset and check that it does what it should do.

Finally, a series of unit tests that checks `read_pose_from_file` `write_pose_to_file` individual function's correctness individually

CI should then tell you the coverage metrics in the codecov job - we need it to be at least 90% to be merged

SteveMacenski commented:
Also as a matter of practice, do not resolve the PR topic discussions. I'll hit "resolve" when I verify that it was completed as intended. It makes my life easier as a maintainer to know that I validated each of my previous comments in your updated fixes without having to go back through all of the collapsed tabs each time :-)

pranavk-2003 commented:
Resolved the conflicts, @SteveMacenski i'll let you know when i want you to check this PR again.

pranavk-2003 commented:
Hey @SteveMacenski its been a while.I have added tests, adjusted the code according to your recommendations and enhancement suggestions.I have tried to solve and fix most of what we discussed last time.Please review the changes and let me know any modifications.Also dont worry about the checks as i will fix them after all the code changes are done.I have also relocated the BaseFootprintPublisher to nav2_toolkit.:)
